search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,433 @@
+                                                      G_VARIANT_TYPE ("s"));
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "content-type",
+                                                      G_VARIANT_TYPE ("s"));
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "thumbnail-uri",
+                                                      G_VARIANT_TYPE ("s"));
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "ekn-id",
+                                                      G_VARIANT_TYPE ("s"));
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "child-tags",
+                                                      G_VARIANT_TYPE ("as"));
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "tags",
+                                                      G_VARIANT_TYPE ("as"));
+
+      g_variant_builder_close (&builder);
+    }
+
+  eks_article_metadata_complete_query (state->provider->skeleton,
+                                       state->invocation,
+                                       (const gchar * const *) shards_strv,
+                                       g_variant_builder_end (&builder));
+}
+
+static GValue *
+append_construction_prop (const gchar *key,
+                          GType        type,
+                          GArray      *values_array,
+                          GPtrArray   *props_array,
+                          guint       *nprops)
+{
+  guint index = values_array->len;
+  g_array_set_size (values_array, values_array->len + 1);
+  g_value_init (&g_array_index (values_array, GValue, index),
+                type);
+  g_ptr_array_add (props_array, g_strdup (key));
+  ++(*nprops);
+
+  return &g_array_index (values_array, GValue, index);
+}
+
+static EkncQueryObject *
+create_query_from_dbus_query_parameters (GVariant    *query_parameters,
+                                         const gchar *application_id)
+{
+  g_autoptr(EkncQueryObject) query = g_object_new (EKNC_TYPE_QUERY_OBJECT, NULL);
+  GVariantIter iter;
+  gchar *iter_key;
+  GVariant *iter_value;
+
+  g_autoptr(GArray) values_array = g_array_new (FALSE, TRUE, sizeof (GValue));
+  g_autoptr(GPtrArray) props_array = g_ptr_array_new_with_free_func (g_free);
+  guint nprops = 0;
+
+  /* Always add the app-id to the query */
+  g_value_set_string (append_construction_prop ("app-id",
```

Comment:
Are these properties all construct-only? If not, I'd probably prefer to avoid marshalling to `GValue` and setting properties directly on the object.

Reply:
They might be. Let me check.

Reply:
Yup, they are, I'm afraid: https://github.com/endlessm/eos-knowledge-lib/blob/master/ekncontent/ekncontent/eknc-query-object.c#L322

Is my solution at least somewhat elegant? ;-) I appreciate that having to do this GValue marshalling dance was a little awkward indeed.